File: rust/parity/specs/eip4337_account_abstraction_sentry.py
"""Parity spec for PiEIP4337AccountAbstractionSentry.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiEIP4337AccountAbstractionSentry"

_mod = load_py_agent("pi_eip4337_account_abstraction_sentry.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiEIP4337AccountAbstractionSentry()
    out = agent.audit_account_abstraction(_mod.AccountAbstractionInput(**data))
    return out.model_dump()


# A clean Smart Account validation function with no forbidden global access.
_SECURE = (
    "contract Account {\n"
    "  function validateUserOp(UserOp calldata op, bytes32 h, uint256 m) "
    "external returns (uint256) {\n"
    "    require(op.signature.length == 65);\n"
    "    return 0;\n"
    "  }\n"
    "}\n"
)

# validateUserOp reaching for tx.origin (forbidden).
_TX_ORIGIN = (
    "function validateUserOp(UserOp op) external returns (uint256) {\n"
    "  require(tx.origin == owner);\n"
    "  return 0;\n"
    "}\n"
)

# Paymaster validation touching block.timestamp (forbidden).
_BLOCK_TIMESTAMP = (
    "function validatePaymasterUserOp(UserOp op, bytes32 h, uint256 m) "
    "external returns (bytes memory, uint256) {\n"
    "  if (block.timestamp > deadline) revert();\n"
    "}\n"
)

# block.number forbidden access.
_BLOCK_NUMBER = (
    "function validateUserOp(UserOp op) external {\n"
    "  uint256 n = block.number;\n"
    "}\n"
)

# gasleft() forbidden access.
_GASLEFT = (
    "function validateUserOp(UserOp op) external {\n"
    "  uint256 g = gasleft();\n"
    "}\n"
)

# block.blockhash forbidden access.
_BLOCKHASH = (
    "function validatePaymasterUserOp(UserOp op) external {\n"
    "  bytes32 b = block.blockhash;\n"
    "}\n"
)

# Multiple forbidden accesses in one validation function -> multiple findings,
# plus a non-validation function that should be ignored entirely.
_MULTI = (
    "contract Mix {\n"
    "  function validateUserOp(UserOp op) external returns (uint256) {\n"
    "    require(tx.origin == owner);\n"
    "    uint256 ts = block.timestamp;\n"
    "    uint256 n = block.number;\n"
    "    return 0;\n"
    "  }\n"
    "  function helper() public { uint256 x = gasleft(); }\n"
    "}\n"
)

# Forbidden access but inside a NON-validation function -> should pass clean.
_FORBIDDEN_NON_VALIDATION = (
    "function transfer(address to) public {\n"
    "  require(tx.origin == msg.sender);\n"
    "  uint256 t = block.timestamp;\n"
    "}\n"
)

SAMPLES = [
    {"input": {"file_path": "Account.sol", "solidity_code": _SECURE}},
    {"input": {"file_path": "Account.sol", "solidity_code": _TX_ORIGIN}},
    {"input": {"file_path": "Paymaster.sol", "solidity_code": _BLOCK_TIMESTAMP}},
    {"input": {"file_path": "Account.sol", "solidity_code": _BLOCK_NUMBER}},
    {"input": {"file_path": "Account.sol", "solidity_code": _GASLEFT}},
    {"input": {"file_path": "Paymaster.sol", "solidity_code": _BLOCKHASH}},
    {"input": {"file_path": "Mix.sol", "solidity_code": _MULTI}},
    {"input": {"file_path": "Token.sol", "solidity_code": _FORBIDDEN_NON_VALIDATION}},
    # empty / edge input
    {"input": {"file_path": "Empty.sol", "solidity_code": ""}},
    {"input": {"file_path": "Account.sol", "solidity_code": _TX_ORIGIN,
               "check_level": "MEDIUM"}},
    # non-strict env -> WARN path, is_secure coerced back to True
    {"input": {"file_path": "Account.sol", "solidity_code": _TX_ORIGIN},
     "env": {"PI_AA_SENTRY_STRICT_MODE": "false"}},
    # explicit strict env -> REJECTED path
    {"input": {"file_path": "Account.sol", "solidity_code": _TX_ORIGIN},
     "env": {"PI_AA_SENTRY_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/eip4337_account_abstraction_sentry.rs
//! Port of `pi_micro_agents/pi_eip4337_account_abstraction_sentry.py`.
//!
//! Audits Solidity Smart Accounts / Paymasters for ERC-4337 bundler simulation
//! restrictions: validation functions (`validateUserOp` /
//! `validatePaymasterUserOp`) must not access forbidden global state such as
//! `tx.origin`, `block.blockhash`, `block.timestamp`, `block.number` or
//! `gasleft()`. Behaviour is a line-for-line mirror of the Python original.

use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub file_path: String,
    pub solidity_code: String,
    #[serde(default = "default_check_level")]
    pub check_level: String,
}

fn default_check_level() -> String {
    "STRICT".to_string()
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub vulnerable_functions: Vec<String>,
    pub flagged_findings: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

// Mirrors: re.findall(r'function\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^{]*\{([\s\S]*?)\}', code)
// Python `.` does not match newlines by default; neither does Rust's. `[\s\S]`
// matches everything (incl. newlines) in both engines. Non-greedy semantics are
// identical between Python `re` and the Rust `regex` crate.
static FUNC_BLOCK_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"function\s+([a-zA-Z0-9_]+)\s*\((.*?)\)[^{]*\{([\s\S]*?)\}").unwrap()
});

// Forbidden global-state patterns, in the exact order the Python list declares
// them. Each tuple is (compiled regex, human-readable keyword).
static FORBIDDEN_PATTERNS: Lazy<Vec<(Regex, &'static str)>> = Lazy::new(|| {
    vec![
        (Regex::new(r"\btx\.origin\b").unwrap(), "tx.origin"),
        (Regex::new(r"\bblock\.blockhash\b").unwrap(), "block.blockhash"),
        (Regex::new(r"\bblock\.timestamp\b").unwrap(), "block.timestamp"),
        (Regex::new(r"\bblock\.number\b").unwrap(), "block.number"),
        (Regex::new(r"\bgasleft\s*\(").unwrap(), "gasleft()"),
    ]
});

/// Mirrors `is_strict_mode()`: strict unless the env var is set to a value that
/// is not (case-insensitively) "true".
///
/// NOTE: the Python original additionally falls back to a
/// `~/.antigravitycli/config.json` (or repo-relative) file when the env var is
/// unset, returning `bool(data.get("PI_AA_SENTRY_STRICT_MODE", True))`. With no
/// such file present the Python default is `True`, which this env-only mirror
/// reproduces exactly. See parity deviations.
fn is_strict_mode() -> bool {
    match std::env::var("PI_AA_SENTRY_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn audit_account_abstraction(input: &Input) -> Output {
    let code = &input.solidity_code;
    let mut vulnerable_funcs: Vec<String> = Vec::new();
    let mut flagged_findings: Vec<String> = Vec::new();

    // Find all functions.
    for caps in FUNC_BLOCK_RE.captures_iter(code) {
        let name = caps.get(1).map(|m| m.as_str()).unwrap_or("");
        // args (group 2) is captured by Python but unused in the loop body.
        let body = caps.get(3).map(|m| m.as_str()).unwrap_or("");

        // Check for paymaster or account abstraction validation methods:
        // validateUserOp, validatePaymasterUserOp
        if name.contains("validateUserOp") || name.contains("validatePaymasterUserOp") {
            // ERC-4337 bans accessing global block metadata, blockhash,
            // gasleft, tx.origin, timestamp, number etc.
            for (pattern, keyword) in FORBIDDEN_PATTERNS.iter() {
                if pattern.is_match(body) {
                    vulnerable_funcs.push(name.to_string());
                    flagged_findings.push(format!(
                        "Validation function '{name}' accesses forbidden global state parameter '{keyword}'. \
This violates ERC-4337 bundler simulation restrictions, causing transaction rejection."
                    ));
                }
            }
        }
    }

    let mut is_secure = vulnerable_funcs.is_empty();
    let risk_score = if !is_secure { 75.0 } else { 0.0 };

    let is_strict = is_strict_mode();
    let mut status = "PASSED".to_string();
    if !is_secure {
        if is_strict {
            status = "REJECTED_AA_RISK".to_string();
        } else {
            status = "WARN_AA_RISK".to_string();
            is_secure = true;
        }
    }

    Output {
        is_secure,
        vulnerable_functions: vulnerable_funcs,
        flagged_findings,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = audit_account_abstraction(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serial_test::serial;

    fn run(code: &str) -> Output {
        audit_account_abstraction(&Input {
            file_path: "f.sol".into(),
            solidity_code: code.into(),
            check_level: "STRICT".into(),
        })
    }

    #[test]
    #[serial]
    fn secure_code_passes() {
        std::env::remove_var("PI_AA_SENTRY_STRICT_MODE");
        let o = run(
            "contract A { function validateUserOp(UserOp op) external returns (uint256) { return 0; } }",
        );
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.vulnerable_functions.is_empty());
    }

    #[test]
    #[serial]
    fn tx_origin_in_validate_rejected_strict() {
        std::env::set_var("PI_AA_SENTRY_STRICT_MODE", "true");
        let o = run(
            "function validateUserOp(UserOp op) external { require(tx.origin == owner); }",
        );
        assert!(!o.is_secure);
        assert_eq!(o.status, "REJECTED_AA_RISK");
        assert_eq!(o.risk_score, 75.0);
        assert_eq!(o.vulnerable_functions, vec!["validateUserOp"]);
        assert_eq!(o.flagged_findings.len(), 1);
    }

    #[test]
    #[serial]
    fn warn_path_coerces_secure_when_not_strict() {
        std::env::set_var("PI_AA_SENTRY_STRICT_MODE", "false");
        let o = run(
            "function validatePaymasterUserOp(UserOp op) external { uint t = block.timestamp; }",
        );
        // is_secure is coerced back to true in non-strict mode
        assert!(o.is_secure);
        assert_eq!(o.status, "WARN_AA_RISK");
        assert_eq!(o.risk_score, 75.0);
        assert_eq!(o.vulnerable_functions, vec!["validatePaymasterUserOp"]);
        std::env::remove_var("PI_AA_SENTRY_STRICT_MODE");
    }

    #[test]
    #[serial]
    fn non_validation_function_ignored() {
        std::env::remove_var("PI_AA_SENTRY_STRICT_MODE");
        let o = run("function harmless() public { uint x = block.number; }");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
    }
}
